### ISSUE TITLE
Incorrect hold timer is used for expiration of srv records from server-state-file

### ISSUE THREAD

**Issue by eglinux** [haproxy/haproxy#2812]
### Detailed Description of the Problem

When using haproxy with dynamic backend servers discovery via dns/srv records and loading from server-state-file the incorrect hold timer is used for setting expiration date of SRV records loaded from the state file.
```hold.obsolete``` has to be used instead of ```hold.timeout``` as according to documentation it controls the expiration of particular srv-record

### Expected Behavior

When haproxy loads SRV records from server-state-file ```hold.obsolete``` timer has to define their expiration time

### Steps to Reproduce the Behavior

1. Run haproxy with loading state file, local resolver and dynamic backend-servers discovery via server-template and srv records. Example of minimal configuration:
```
global
  chroot  /usr/local/haproxy-3.0.6
  user  haproxy
  group  haproxy
  server-state-file  /usr/local/haproxy-3.0.6/status
  stats  socket /usr/local/haproxy-3.0.6/stats level admin user root group root mode 0600

defaults
  load-server-state-from-file  global
  mode http
  timeout client 1m
  timeout server 1m
  timeout connect 10s

resolvers local-test
  nameserver ns01 127.0.0.1:5353
  hold valid 5s
  hold nx 5d
  hold refused 5d
  hold timeout 5d
  hold other 5d
  hold obsolete 0
  accepted_payload_size 8192

frontend a
  bind 127.0.0.1:5580
  use_backend b

backend b
  balance roundrobin
  server-template b-local 2 _back._tcp.local.test resolvers local-test init-addr last,libc,none resolve-opts allow-dup-ip resolve-prefer ipv4 check
  option httpchk GET /healthcheck
  http-check expect status 200
  ```
  2. Run local named with 2 srv records in local.test zone. Example of zone file:
  ```
@ 1800 IN SOA ns01.local.test. hostmaster.local.test. (
                2024091201      ; serial number
                10000           ; refresh
                3600            ; retry
                1209600         ; expire
                1800 )  ; Negative Cach TTL

@                       IN      NS      ns01.local.test.

localhost               IN      A       127.0.0.1
ns01			IN	A	127.0.0.1
_back._tcp.local.test. SRV 10 100 5355 localhost.local.test.
_back._tcp.local.test. SRV 10 100 5356 localhost.local.test.
```
3. Dump haproxy state to filesystem.
```
[#] echo 'show servers state' | nc -U stats > status && cat status
1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
3 b 1 b-local1 127.0.0.1 2 0 1 1 19 15 3 4 6 0 0 0 localhost.local.test 5356 _back._tcp.local.test 0 0 - - 0
3 b 2 b-local2 127.0.0.1 2 0 1 1 18 15 3 4 6 0 0 0 localhost.local.test 5355 _back._tcp.local.test 0 0 - - 0
```
4. Remove one srv record from named:
```
[#] dig srv _back._tcp.local.test. @127.0.0.1 -p5353 +short
10 100 5355 localhost.local.test.
```
Haproxy successfully detects the change and removes it from its state:
```
[#] echo 'show servers state' | nc -U stats
1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
3 b 1 b-local1 - 0 96 1 1 1 15 3 0 14 0 0 0 - 0 _back._tcp.local.test 0 0 - - 0
3 b 2 b-local2 127.0.0.1 2 64 1 1 11 15 3 4 6 0 0 0 localhost.local.test 5355 _back._tcp.local.test 0 0 - - 0
```
5. Reload haproxy. It will load servers from state file, but not from dns view:
```
[#] echo 'show servers state' | nc -U stats
1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
3 b 1 b-local1 127.0.0.1 2 0 1 1 19 15 3 4 6 0 0 0 localhost.local.test 5356 _back._tcp.local.test 0 0 - - 0
3 b 2 b-local2 127.0.0.1 2 0 1 1 18 15 3 4 6 0 0 0 localhost.local.test 5355 _back._tcp.local.test 0 0 - - 0
```

In the above configuration ```hold.timeout``` is set to ```5d```, so the stale record which is not returned by dns server anymore will expire only after 5d.

### Do you have any idea what may have caused this?

It looks like when srv records are loaded from state file the expiration time is controlled by ```hold.timeout``` instead of ```hold.obsolete```

https://github.com/haproxy/haproxy/blob/master/src/server_state.c#L429
```
task_schedule(srv->srvrq_check, tick_add(now_ms, srv->srvrq->resolvers->hold.timeout));
```

### Do you have an idea how to solve the issue?

The simple patch can change that to ```hold.obsolete```

```
From: Andrii Ustymeno <andrii.ustymenko@adyen.com>
Date: Fir, 06 Dec 2024 15:28:37 +0100
Subject: BUG/MINOR: server_state: use hold.obsolete timer for initial expiration time of srv records

After loading srv records from server-state-file use hold.obsolete timer for the expiration time of particular record

---
 src/server_state.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/server_state.c b/src/server_state.c
index 131f339ea..f36f8e1c7 100644
--- a/src/server_state.c
+++ b/src/server_state.c
@@ -426,7 +426,7 @@ static void srv_state_srv_update(struct server *srv, int version, char **params)

                /* insert in tree and set the srvrq expiration date */
                ebis_insert(&srv->srvrq->named_servers, &srv->host_dn);
-               task_schedule(srv->srvrq_check, tick_add(now_ms, srv->srvrq->resolvers->hold.timeout));
+               task_schedule(srv->srvrq_check, tick_add(now_ms, srv->srvrq->resolvers->hold.obsolete));

                /* Unset SRV_F_MAPPORTS for SRV records.
                 * SRV_F_MAPPORTS is unfortunately set by parse_server()
```

### What is your configuration?

```haproxy
global
  chroot  /usr/local/haproxy-3.0.6
  user  haproxy
  group  haproxy
  server-state-file  /usr/local/haproxy-3.0.6/status
  stats  socket /usr/local/haproxy-3.0.6/stats level admin user root group root mode 0600

defaults
  load-server-state-from-file  global
  mode http
  timeout client 1m
  timeout server 1m
  timeout connect 10s

resolvers local-test
  nameserver ns01 127.0.0.1:5353
  hold valid 5s
  hold nx 5d
  hold refused 5d
  hold timeout 5d
  hold other 5d
  hold obsolete 0
  accepted_payload_size 8192

frontend a
  bind 127.0.0.1:5580
  use_backend b

backend b
  balance roundrobin
  server-template b-local 2 _back._tcp.local.test resolvers local-test init-addr last,libc,none resolve-opts allow-dup-ip resolve-prefer ipv4 check
  option httpchk GET /healthcheck
  http-check expect status 200
```

### Output of `haproxy -vv`

```plain
HAProxy version 3.0.6-c2c0090 2024/11/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2029.
Known bugs: http://www.haproxy.org/bugs/bugs-3.0.6.html
Running on: Linux 5.14.0-427.35.1.el9_4.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Sep 11 12:23:45 PDT 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv -DMAX_SESS_STKCTR=12
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1 USE_PCRE2=1
  DEBUG   =

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=4).
Built with OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
Running on OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.5.0 20240719 (Red Hat 11.5.0-2.0.1)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

### Additional Information

In our environment we have ```hold.timeout``` set to very high value and ```hold.obsolete``` to 0. The intention is to have always fresh dns view, but also to be protected against dns-servers issues if they occur. So, the described issue has two negative impacts on us:

1. When haproxy crashes and gets restarted by systemd it can pickup outdated state-file 
2. When reloading haproxy we have a hook to dump state-file before stop, but it can cause a race-condition with dns-view being changed right after that, but before new worker actually forked. 

This issue detected on all release branches

**Comment by capflam** [comment]
You're right, there is an issue in your exemple. However, your fix is not the right one. Especially with an `obsolete` timeout set to 0. In your case, after a reload, all your servers will appear immediately as `DOWN` because the `srvrq_check` task of each server will expire before the first resolution. A good way to observe that is to update your server-state file, stop your DNS server and reload haproxy. In that case, your servers should inherit their state from the file, as expected, and remain in this state for `hold.timeout`  milliseconds, the resolution timeout. To be strict, it should even be `resolve_retries * hold.timeout` milliseconds. With your fix, servers will be set to `DOWN` state for maintenance. It is not the expected behavior.

Unfortunately, the fix is not so trivial. The cleanup must be performed when the DNS response is processed, in `resolv_check_response()` function. Servers not attached to the `SRVRQ` resolution  must be cleaned up after the `obsolete` timeout.

**Comment by eglinux** [comment]
So, the same behavior with ```DOWN``` state would happen if ```hold.timeout``` is set to 0 which probably also not desired.

As I understood from the code, this case cannot be covered by ```resolv_check_response()``` because in that function the different structure is being checked, which contains the dns responses, so basically the dns_view. And entries appear there only after they have been returned at least once by the dns server. Records checked and evicted by ```last_seen``` field which gets updated every successful dns request. So, to make it working there the records from state file have to be copied to that structure first on server start. And state file can contain both static and dynamic records which might make it another issue.

And it also makes sense that state file might be not 100% correct and contain stale records. It is basically best effort approach.
So, basically scheduling the task to update it with dns_view for srv records particularly after start I think the best haproxy can do in this case. If ```obsolete``` is bad timer for that because it is 0 in most of the cases and ```timeout``` is quite high, maybe it makes sense to introduce another hold parameter which should be >0, but relatively small? Like in my case ```hold.valid``` would also work, which would mean that the stale records will be evicted in 5s after boot. Semantically speaking it is also correct, depending how the records from state file are being treated either ```valid``` or ```obsolete```. I guess it makes sense that ```hold.valid``` is never 0.

**Comment by capflam** [comment]
Well, I think I misunderstood the `hold.timeout` purpose. The idea was to affect the resolution timeout to `srvrq_check` tasks. But in fact, `hold.timeout` is not the resolution timeout. It is the delay to considered a valid DNS items as expired. It is evaluated after a resolution timeout was triggered. `hold.*` timeouts are in fact related to response items. So none of these timeouts are applicable in your case. But the original idea remains valid. The resolution timeout must be used. Because it is the idea. The server's state inherited from the state-file should be considered as outdated if no resolution renew it.
To sum up, I guess it should be `timeout.resolve + resolve_retries * timeout.retry` instead of `hold.timeout`. The default value will be 4s (1s + 3 * 1s).

Here is the patch:
```diff
diff --git a/src/server_state.c b/src/server_state.c
index 131f339ea..df6c32bd2 100644
--- a/src/server_state.c
+++ b/src/server_state.c
@@ -426,7 +426,9 @@ static void srv_state_srv_update(struct server *srv, int version, char **params)
 
                /* insert in tree and set the srvrq expiration date */
                ebis_insert(&srv->srvrq->named_servers, &srv->host_dn);
-               task_schedule(srv->srvrq_check, tick_add(now_ms, srv->srvrq->resolvers->hold.timeout));
+               task_schedule(srv->srvrq_check, tick_add(now_ms, srv->srvrq->resolvers->timeout.resolve +
+                                                        srv->srvrq->resolvers->resolve_retries *
+                                                        srv->srvrq->resolvers->timeout.retry));
 
                /* Unset SRV_F_MAPPORTS for SRV records.
                 * SRV_F_MAPPORTS is unfortunately set by parse_server()
```

what do you think of this version?

**Comment by eglinux** [comment]
Yeah, that makes more sense. And there will be no cases when it can end up as 0. It also fixes the issue for us, so I would vote both hands for this patch.

**Comment by capflam** [comment]
Copy that. I will push the fix tomorrow morning. Thanks !

**Comment by capflam** [comment]
It should be fixed now. But the wrong issue was referenced in the commit. For records, the commit is 647a2906626e9c3d9c3349d338a35798325496f2
